src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutionParser.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.Metadata;
+using Internal.TypeSystem;
+using System.Xml.XPath;
+using System.Globalization;
+
+namespace ILCompiler
+{
+    internal class SubstitutionsParser : ProcessLinkerXmlBase
+    {
+        private readonly Dictionary<MethodDesc, BodySubstitution> _methodSubstitutions;
+        private readonly Dictionary<FieldDesc, object> _fieldSubstitutions;
+
+        private SubstitutionsParser(TypeSystemContext context, UnmanagedMemoryStream documentStream, ManifestResource resource, ModuleDesc resourceAssembly, string xmlDocumentLocation, IReadOnlyDictionary<string, bool> featureSwitchValues)
+                : base(context, documentStream, resource, resourceAssembly, xmlDocumentLocation, featureSwitchValues)
+        {
+            _methodSubstitutions = new Dictionary<MethodDesc, BodySubstitution>();
+            _fieldSubstitutions = new Dictionary<FieldDesc, object>();
+        }
+
+        protected override void ProcessAssembly(ModuleDesc assembly, XPathNavigator nav, bool warnOnUnresolvedTypes)
+        {
+            ProcessTypes(assembly, nav, warnOnUnresolvedTypes);
+        }
+
+        // protected override TypeDesc? ProcessExportedType(ExportedType exported, ModuleDesc assembly, XPathNavigator nav) => null;
+
+        protected override bool ProcessTypePattern(string fullname, ModuleDesc assembly, XPathNavigator nav) => false;
+
+        protected override void ProcessType(TypeDesc type, XPathNavigator nav)
+        {
+            Debug.Assert(ShouldProcessElement(nav));
+            ProcessTypeChildren(type, nav);
+        }
+
+        protected override void ProcessMethod(TypeDesc type, MethodDesc method, XPathNavigator nav, object customData)
```

Comment:
Could you do the thing again where the first commit is just reformatted source code from the linker repo and the next commit is this?

Looks like there's subtle differences that would be easier to spot in the diff.

On the linker side, we actually override the `(TypeDefinition type, XPathNavigator methodNav, object? _customData)` overload, not this one, and that causes a behavioral difference:

```xml
<method name="IsEnabled" body="stub" />
```

Is not valid in il linker, but this code would generate a substitution for all overloads of IsEnabled.

```xml
<method name="IsEnabled" signature="System.Int32 GetIntConstant()" body="stub" />
```

Would create a substitution for GetIntConstant and ignore IsEnabled in IL Linker, but this code would substitute GetIntConstant and all overloads of IsEnabled.

The reason why we're switching to IL Linker's parser is to never have to worry about these unspecified behaviors of the linker format.